### PR TITLE
Refining AudioWorklet introduction

### PR DESCRIPTION
**Pull request by hoch**
PTAL - @joeberkovitz, @rtoy, and @padenot.

The branch for preview: http://hoch.github.io/web-audio-api/#AudioWorklet

- Added an informative diagram.
- Added a simple code example.
- Rearranged the text to provide the conceptual overview.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -5071,40 +5071,87 @@ <h2 id="AudioWorkerParamDescriptor">
...
+            synchronous execution of the script code with other built-in
+            <code>AudioNode</code>s in the audio graph.
+          </p>
+          <p>
+            An associated pair of objects must be defined in order to realize
+            this mechanism; <code>AudioWorkletNode</code> and
+            <code>AudioWorkletProcessor</code>. The former represents the
+            interface for the main global scope similar to other AudioNode
+            objects and the latter implements the internal audio processing
+            within a special scope named <code>AudioWorkletGlobaScope</code>.
```
Comment (joeberkovitz) [inline index.html]: sp - AudioWorkletGloba*l*Scope

File: index.html [hunk 2]
```diff
@@ -5071,40 +5071,87 @@ <h2 id="AudioWorkerParamDescriptor">
...
+          </figure>
+          <blockquote>
+            <pre>// In the main global scope
+window.audioWorklet.import("bypass.js").then(function () {
+  var context = new AudioContext();
+  var bypass = new AudioWorkletNode(context, "Bypass");
+});</pre>
+          </blockquote>
+          <p>
+            An <code>AudioWorkletNode</code> and its corresponding processor
```
Comment (joeberkovitz) [inline index.html]: I thought that only the Processor could be registered in the imported file. The Node itself is never actually referenced in the imported JS since it's all within the AWGS.

File: index.html [hunk 3]
```diff
@@ -5071,40 +5071,87 @@ <h2 id="AudioWorkerParamDescriptor">
...
+            can be registered by importing a JavaScript file containing a
+            definition of <code>AudioWorkletProcessor</code>. It is stored in
+            the <em>node name to node definition map</em> which belongs to the
```
Comment (joeberkovitz) [inline index.html]: Likewise perhaps this should be called _node name to processor map_

**Comment by joeberkovitz** [comment]
I think the diagram would be more informative if it showed some representation of audio input and output going into and out of the Processor.

**Comment by joeberkovitz** [comment]
Sorry, mistaken click

File: index.html (hunk 3 above)
Comment (hoch) [inline index.html]: I think we need several maps for different purposes. You'll see them in the PRs over next few days.

- node name to node definition map in the main global scope
- node name to processor definition map in AWGS
- node instance map in AudioContext

File: index.html (hunk 2 above)
Comment (hoch) [inline index.html]: I'll rephrase this. It was a mistake.

File: index.html [hunk 4]
```diff
@@ -5071,40 +5071,87 @@ <h2 id="AudioWorkerParamDescriptor">
...
+          <p>
+            The <code>AudioWorklet</code> object allows developers to supply
+            JavaScript to process audio on the <a>rendering thread</a>,
+            supporting custom AudioNodes. This processing mechanism ensures the
+            synchronous execution of the script code with other built-in
+            <code>AudioNode</code>s in the audio graph.
+          </p>
+          <p>
+            An associated pair of objects must be defined in order to realize
+            this mechanism; <code>AudioWorkletNode</code> and
```
Comment (rtoy) [inline index.html]: Grammar: "mechaniscm;" -> "mechanism:" (use colon).

File: index.html [hunk 5]
```diff
@@ -5071,40 +5071,87 @@ <h2 id="AudioWorkerParamDescriptor">
...
+          <figure>
+            <img alt="AudioWorklet concept" src=
+            "images/audioworklet-concept.png">
+            <figcaption>
+              <code>AudioWorkletNode</code> and
+              <code>AudioWorkletProcessor</code>
+            </figcaption>
+          </figure>
+          <blockquote>
+            <pre>// In the main global scope
```
Comment (rtoy) [inline index.html]: Can you put line breaks after pre and before /pre, like we do for other html tags?

File: index.html (hunk 3 above)
Comment (rtoy) [inline index.html]: What's the point of emphasizing the phrase? Are you going to define these later?  If so, please make it a link.

File: index.html [hunk 6]
```diff
@@ -5071,40 +5071,87 @@ <h2 id="AudioWorkerParamDescriptor">
...
+registerProcessor("Bypass", class extends AudioWorkletProcessor {
+  process (inputs, outputs) {
+    // Single input, single channel.
+    var input = inputs[0], output = outputs[0];
+    output[0].set(input[0]);
+  }
+});</pre>
+          </blockquote>
+          <p>
+            At the instantiation of <code>AudioWorkletNode</code> in the main
```
Comment (rtoy) [inline index.html]: "main" is redundant?

File: index.html [hunk 7]
```diff
@@ -5071,40 +5071,87 @@ <h2 id="AudioWorkerParamDescriptor">
...
+            global scope, the counterpart <code>AudioWorkletProcessor</code>
+            will also be created in <code>AudioWorkletGlobalScope</code>. These
+            two objects communicate via the asynchronous message passing
+            described in the processing model section.
```
Comment (rtoy) [inline index.html]: Need a link for the processing model section.

File: index.html [hunk 8]
```diff
@@ -5071,40 +5071,87 @@ <h2 id="AudioWorkerParamDescriptor">
...
+          </p>
+        </section>
+        <dl title="partial interface Window" class="idl">
+          <dt>
+            [SameObject] readonly attribute Worklet audioWorklet
+          </dt>
+          <dd>
+            <p>
+              The <code>audioWorklet</code> attribute allows access to the
+              <code>Worklet</code> responsible to the classes such as
```
Comment (rtoy) [inline index.html]: What does "responsible to" mean here?

File: index.html (hunk 3 above)
Comment (hoch) [inline index.html]: I am following the formatting from https://www.w3.org/TR/css-paint-api-1/. Not sure how to apply it to the respec. Will making it a link have the effect? This is not a part of the IDL, so not sure how this can be formally defined with a link.

File: index.html (hunk 6 above)
Comment (hoch) [inline index.html]: I don't think so. We have different types of global scope, but I think we can use `window` instead.

File: index.html (hunk 7 above)
Comment (hoch) [inline index.html]: The new message passing mechanism is not defined by the processing model yet. I can remove the sentence until we define the algorithm.

File: index.html (hunk 8 above)
Comment (hoch) [inline index.html]: I can rephrase this to:
The audioWorklet attributes allows access to the Worklet object that can import a JavaScript file containing AudioWorkletProcessor class definitions.

**Comment by hoch** [comment]
Is there anything more to do here?

**Comment by joeberkovitz** [comment]
I feel that the label "getter and setter methods" in the diagram seems a little overspecific -- many nodes have useful methods that are neither getters nor setters, as well as AudioParams (which are neither). A label that makes a good counterpart to "audio in and audio out" (for the Processor) might be something like "application-level interaction"

**Comment by joeberkovitz** [comment]
I like it!

**Comment by hoch** [comment]
Can you merge this section? The PR for next section is also waiting in line. :)

**Comment by joeberkovitz** [comment]
Merged.

**Comment by hoch** [comment]
Thanks!
